"Task not serializable" when writing to S3 layer

This is issue against master. When trying to write a layer to S3 with:

```scala
val pyramid = Pyramid.fromLayerRDD(layer, resampleMethod = Sum)
val store = AttributeStore(catalog)
val writer = LayerWriter(store, catalog)
pyramid.write(layerName, writer, ZCurveKeyIndexMethod)
```

The following exception will be thrown:

```
geotrellis.store.package$LayerWriteError: Failed to write Layer(name = "MWI", zoom = 6)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:125)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:49)
	at geotrellis.spark.store.LayerWriter$class.write(LayerWriter.scala:150)
	at geotrellis.spark.store.s3.S3LayerWriter.write(S3LayerWriter.scala:49)
	at geotrellis.spark.pyramid.Pyramid$$anonfun$write$1.apply$mcVI$sp(Pyramid.scala:68)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
	at geotrellis.spark.pyramid.Pyramid.write(Pyramid.scala:67)
	at geotrellis.sdg.OutputPyramid$.saveLayer(OutputPyramid.scala:53)
	at geotrellis.sdg.PopulationNearRoadsJob$$anonfun$9$$anonfun$apply$11.apply(PopulationNearRoadsJob.scala:171)
	at geotrellis.sdg.PopulationNearRoadsJob$$anonfun$9$$anonfun$apply$11.apply(PopulationNearRoadsJob.scala:148)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: org.apache.spark.SparkException: Task not serializable
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:403)
	at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:393)
	at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:162)
	at org.apache.spark.SparkContext.clean(SparkContext.scala:2326)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:934)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:933)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:933)
	at geotrellis.spark.store.s3.S3RDDWriter.update(S3RDDWriter.scala:78)
	at geotrellis.spark.store.s3.S3RDDWriter.write(S3RDDWriter.scala:51)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:129)
	... 16 more
Caused by: java.io.NotSerializableException: geotrellis.spark.store.s3.S3SparkLayerProvider
Serialization stack:
	- object not serializable (class: geotrellis.spark.store.s3.S3SparkLayerProvider, value: geotrellis.spark.store.s3.S3SparkLayerProvider@67eb115a)
	- field (class: geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, name: $outer, type: class geotrellis.spark.store.s3.S3SparkLayerProvider)
	- object (class geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, <function0>)
	- field (class: geotrellis.spark.store.s3.S3RDDWriter, name: geotrellis$spark$store$s3$S3RDDWriter$$s3Client, type: interface scala.Function0)
	- object (class geotrellis.spark.store.s3.S3RDDWriter, geotrellis.spark.store.s3.S3RDDWriter@75503aa7)
	- field (class: geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, name: $outer, type: class geotrellis.spark.store.s3.S3RDDWriter)
	- object (class geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, <function1>)
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:46)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:400)
	... 28 more
19/09/23 13:42:57 INFO ApplicationMaster: Final app status: FAILED, exitCode: 15, (reason: User class threw exception: geotrellis.store.package$LayerWriteError: Failed to write Layer(name = "MWI", zoom = 6)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:125)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:49)
	at geotrellis.spark.store.LayerWriter$class.write(LayerWriter.scala:150)
	at geotrellis.spark.store.s3.S3LayerWriter.write(S3LayerWriter.scala:49)
	at geotrellis.spark.pyramid.Pyramid$$anonfun$write$1.apply$mcVI$sp(Pyramid.scala:68)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
	at geotrellis.spark.pyramid.Pyramid.write(Pyramid.scala:67)
	at geotrellis.sdg.OutputPyramid$.saveLayer(OutputPyramid.scala:53)
	at geotrellis.sdg.PopulationNearRoadsJob$$anonfun$9$$anonfun$apply$11.apply(PopulationNearRoadsJob.scala:171)
	at geotrellis.sdg.PopulationNearRoadsJob$$anonfun$9$$anonfun$apply$11.apply(PopulationNearRoadsJob.scala:148)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: org.apache.spark.SparkException: Task not serializable
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:403)
	at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:393)
	at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:162)
	at org.apache.spark.SparkContext.clean(SparkContext.scala:2326)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:934)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:933)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:933)
	at geotrellis.spark.store.s3.S3RDDWriter.update(S3RDDWriter.scala:78)
	at geotrellis.spark.store.s3.S3RDDWriter.write(S3RDDWriter.scala:51)
	at geotrellis.spark.store.s3.S3LayerWriter._write(S3LayerWriter.scala:129)
	... 16 more
Caused by: java.io.NotSerializableException: geotrellis.spark.store.s3.S3SparkLayerProvider
Serialization stack:
	- object not serializable (class: geotrellis.spark.store.s3.S3SparkLayerProvider, value: geotrellis.spark.store.s3.S3SparkLayerProvider@67eb115a)
	- field (class: geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, name: $outer, type: class geotrellis.spark.store.s3.S3SparkLayerProvider)
	- object (class geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, <function0>)
	- field (class: geotrellis.spark.store.s3.S3RDDWriter, name: geotrellis$spark$store$s3$S3RDDWriter$$s3Client, type: interface scala.Function0)
	- object (class geotrellis.spark.store.s3.S3RDDWriter, geotrellis.spark.store.s3.S3RDDWriter@75503aa7)
	- field (class: geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, name: $outer, type: class geotrellis.spark.store.s3.S3RDDWriter)
	- object (class geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, <function1>)
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:46)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:400)
	... 28 more
```

For clarity the serialization stack is:
```
- object not serializable (class: geotrellis.spark.store.s3.S3SparkLayerProvider, value: geotrellis.spark.store.s3.S3SparkLayerProvider@67eb115a)
- field (class: geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, name: $outer, type: class geotrellis.spark.store.s3.S3SparkLayerProvider)
- object (class geotrellis.spark.store.s3.S3SparkLayerProvider$$anonfun$layerWriter$1, <function0>)
- field (class: geotrellis.spark.store.s3.S3RDDWriter, name: geotrellis$spark$store$s3$S3RDDWriter$$s3Client, type: interface scala.Function0)
- object (class geotrellis.spark.store.s3.S3RDDWriter, geotrellis.spark.store.s3.S3RDDWriter@75503aa7)
- field (class: geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, name: $outer, type: class geotrellis.spark.store.s3.S3RDDWriter)
- object (class geotrellis.spark.store.s3.S3RDDWriter$$anonfun$update$1, <function1>)
	at 
```

The first problem is that `s3Client` caused a serialization.
I'm unclear how `S3SparkLayerProvider` gets pulled into this serialization chain however.

--- a/s3-spark/src/main/scala/geotrellis/spark/store/s3/S3SparkLayerProvider.scala
+++ b/s3-spark/src/main/scala/geotrellis/spark/store/s3/S3SparkLayerProvider.scala
@@ -32,12 +32,12 @@ import java.net.URI
  */
 class S3SparkLayerProvider extends S3CollectionLayerProvider with LayerReaderProvider with LayerWriterProvider {
   def layerReader(uri: URI, store: AttributeStore, sc: SparkContext): FilteringLayerReader[LayerId] = {
-    new S3LayerReader(store, s3Client)(sc)
+    new S3LayerReader(store, S3ClientProducer.get())(sc)
   }
 
   def layerWriter(uri: URI, store: AttributeStore): LayerWriter[LayerId] = {
     // TODO: encoder ACL changes in putObjectModifier
     val s3Uri = new AmazonS3URI(uri)
-    new S3LayerWriter(store, bucket = s3Uri.getBucket(), keyPrefix = s3Uri.getKey(), identity, s3Client)
+    new S3LayerWriter(store, bucket = s3Uri.getBucket(), keyPrefix = s3Uri.getKey(), identity, S3ClientProducer.get())
   }
 }

--- a/s3/src/main/scala/geotrellis/store/s3/S3CollectionLayerProvider.scala
+++ b/s3/src/main/scala/geotrellis/store/s3/S3CollectionLayerProvider.scala
@@ -26,12 +26,7 @@ import java.net.URI
  * The uri represents S3 bucket an prefix of catalog root.
  *  ex: `s3://<bucket>/<prefix-to-catalog>`
  */
-class S3CollectionLayerProvider()
-  extends AttributeStoreProvider
-     with ValueReaderProvider
-     with CollectionLayerReaderProvider {
-
-  @transient lazy val s3Client = S3ClientProducer.get()
+class S3CollectionLayerProvider() extends AttributeStoreProvider with ValueReaderProvider with CollectionLayerReaderProvider {
 
   def canProcess(uri: URI): Boolean = uri.getScheme match {
     case str: String => if (str.toLowerCase == "s3") true else false
@@ -45,11 +40,11 @@ class S3CollectionLayerProvider()
         case Some(s) => s
         case None => ""
       }
-    new S3AttributeStore(bucket = s3Uri.getBucket(), prefix = prefix, s3Client)
+    new S3AttributeStore(bucket = s3Uri.getBucket(), prefix = prefix, S3ClientProducer.get())
   }
 
   def valueReader(uri: URI, store: AttributeStore): ValueReader[LayerId] = {
-    new S3ValueReader(store, s3Client)
+    new S3ValueReader(store, S3ClientProducer.get())
   }
 
   def collectionLayerReader(uri: URI, store: AttributeStore): CollectionLayerReader[LayerId] = {

